apiserver/storageprovisioner/internal/filesystemwatcher/mock_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package filesystemwatcher_test
+
+import (
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/watcher/watchertest"
+)
+
+type mockBackend struct {
+	machineFilesystemsW           *watchertest.StringsWatcher
+	machineFilesystemAttachmentsW *watchertest.StringsWatcher
+	modelFilesystemsW             *watchertest.StringsWatcher
+	modelFilesystemAttachmentsW   *watchertest.StringsWatcher
+	modelVolumeAttachmentsW       *watchertest.StringsWatcher
+
+	filesystems       map[string]*mockFilesystem
+	volumeAttachments map[string]*mockVolumeAttachment
+}
+
+func (b *mockBackend) Filesystem(tag names.FilesystemTag) (state.Filesystem, error) {
+	if f, ok := b.filesystems[tag.Id()]; ok {
+		return f, nil
+	}
+	return nil, errors.NotFoundf("filesystem %s", tag.Id())
+}
+
+func (b *mockBackend) VolumeAttachment(m names.MachineTag, v names.VolumeTag) (state.VolumeAttachment, error) {
+	if m.Id() != "0" {
+		// The tests all operate on machine "0", and the watchers
+		// should ignore attachments for other machines, so we should
+		// never get here.
+		panic("should not get here")
```

Comment:
Instead of panicing, why not return an error and let the caller's c.Assert(err, jc.ErrorIsNil) fail the test. IMO since the panic would result from a logic error ("code should do such and such"), any such violation is a test failure.